<a href="https://colab.research.google.com/github/JeffersonBP/Aula-_JavascriptJeff/blob/main/BIG_A3_Classificacao.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

<head>
  <meta name="author" content="Rogério de Oliveira">
  <meta institution="author" content="Universidade Presbiteriana Mackenzie">
</head>

<img src="http://meusite.mackenzie.br/rogerio/mackenzie_logo/UPM.2_horizontal_vermelho.jpg" width=300, align="right">
<!-- <h1 align=left><font size = 6, style="color:rgb(200,0,0)"> optional title </font></h1> -->


# **Tarefa de Aprofundamento 3**
## CASO: Controle de Qualidade de Peças de uma Indústria
---
<br>
<br>





A partir deste notebook você vai criar e avaliar diferentes modelos de classificação para prever a qualidade de peças de uma indústria.

Em seguida, com base nos seus resultados, responda o **questionário do Moodle**. (**dica**: o questionário é de alternativas e as respontas irão ajudá-lo a verificar os seus resultados ao longo do programa).

# Basic imports

In [131]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
%matplotlib inline
import seaborn as sns
from sklearn.preprocessing import StandardScaler
import pandas as pd



from sklearn.neighbors import KNeighborsClassifier
from sklearn.model_selection import GridSearchCV
from sklearn.metrics import classification_report



from sklearn.linear_model import LogisticRegression




# Exploração dos Dados



`pieces` é um data set para o controle de qualidade de peças produzidas por uma indústria. São empregadas quatro medidas (A, B, C e D) para o controle da qualidade das peças. As peças são então *Accept, Refurbish* ou *Reject* segundo o controle de qualidade (atributo `Quality`). A indústria ainda conta com 3 unidades para a produção dessas peças (atributo `Unit`).

**Objetivo:** Seu objetivo é criar modelos supervisionados de classificação para predição da qualidade das peças e empregar o melhor modelo obtido (maior acuracidade) na predição da qualidade de novos casos.



In [132]:
df = pd.read_csv('http://meusite.mackenzie.br/rogerio/TIC/pieces1.csv')
df.head()

id    A    B    C    D    Quality Unit
0  559  4.9  3.1  1.5  0.1     Reject   SP
1  629  4.8  3.4  1.6  0.2     Reject   SP
2  192  6.7  2.5  5.8  1.8  Refurbish   RJ
3  359  7.6  3.0  6.6  2.1  Refurbish   RJ
4    9  4.9  3.1  1.5  0.1     Reject   RJ

## Inspecione os dados

In [133]:
# seu código
# imprimir as primeiras 5 linhas do dataframe

print("Primeiras 5 linhas do data frame")
print(df.head())

# Imprimir as últimas 5 linhas do DataFrame
print("Ultimas  5 linhas do data frame")
print(df.tail())

# Verificar informações básicas sobre os dados
print("\nInformações básicas sobre os dados:")
print(df.info())

# Verificar estatísticas descritivas
print("\nEstatísticas descritivas:")
print(df.describe())

# Lidar com valores nulos (se houver)
# Remover linhas com valores nulos
df = df.dropna()

# Verificar se há valores nulos restantes
print("\nVerificar se há valores nulos:")
print(df.isnull().sum())
# Visualização dF
print("\nVisualização:")
display(df)

Primeiras 5 linhas do data frame
    id    A    B    C    D    Quality Unit
0  559  4.9  3.1  1.5  0.1     Reject   SP
1  629  4.8  3.4  1.6  0.2     Reject   SP
2  192  6.7  2.5  5.8  1.8  Refurbish   RJ
3  359  7.6  3.0  6.6  2.1  Refurbish   RJ
4    9  4.9  3.1  1.5  0.1     Reject   RJ
Ultimas  5 linhas do data frame
      id    A    B    C    D    Quality Unit
495  155  4.6  3.6  1.0  0.2     Reject   SP
496  120  6.9  3.2  5.7  2.3  Refurbish   RJ
497  321  6.7  3.1  5.6  2.4  Refurbish   RJ
498  448  5.6  3.0  4.1  1.3     Accept   SP
499  290  4.4  2.9  1.4  0.2     Reject   BH

Informações básicas sobre os dados:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 500 entries, 0 to 499
Data columns (total 7 columns):
 #   Column   Non-Null Count  Dtype  
---  ------   --------------  -----  
 0   id       500 non-null    int64  
 1   A        500 non-null    float64
 2   B        491 non-null    float64
 3   C        494 non-null    float64
 4   D        496 non-null    float64


id    A    B    C    D    Quality Unit
0    559  4.9  3.1  1.5  0.1     Reject   SP
1    629  4.8  3.4  1.6  0.2     Reject   SP
2    192  6.7  2.5  5.8  1.8  Refurbish   RJ
3    359  7.6  3.0  6.6  2.1  Refurbish   RJ
4      9  4.9  3.1  1.5  0.1     Reject   RJ
..   ...  ...  ...  ...  ...        ...  ...
495  155  4.6  3.6  1.0  0.2     Reject   SP
496  120  6.9  3.2  5.7  2.3  Refurbish   RJ
497  321  6.7  3.1  5.6  2.4  Refurbish   RJ
498  448  5.6  3.0  4.1  1.3     Accept   SP
499  290  4.4  2.9  1.4  0.2     Reject   BH

[481 rows x 7 columns]

## Preparação dos dados

Aqui você vai preparar o dado para o uso nos modelos. Você fará a seleção de atributos, tratamento de valores nulos, o encode dos dados, normalização e a separação dos dados de treinamento e teste.

### Seleção de Atributos

Elimine atributos que não fazem parte do treinamento.

In [134]:
# seu código

# Lidar com valores nulos (se houver)
# Remover linhas com valores nulos
df = df.dropna()



### Tratamento de Dados Ausentes

Verifique as presença de valores ausentes e, se houverem, subtitua pelo valor médio dos atributos.

In [135]:
# seu código


# Verifica se há valores ausentes
print("\nVerificar se há valores nulos:")
print(df.isnull().sum())

# Substitui os valores ausentes pelo valor médio dos atributos
for col in df.columns:
    if df[col].isnull().any():
        df[col].fillna(df[col].mean(), inplace=True)

# Verifica se há valores ausentes
print("\nVerificar se há valores nulos:")
print(df.isnull().sum())



Verificar se há valores nulos:
id         0
A          0
B          0
C          0
D          0
Quality    0
Unit       0
dtype: int64

Verificar se há valores nulos:
id         0
A          0
B          0
C          0
D          0
Quality    0
Unit       0
dtype: int64


### Hot encode

Faça os *Hot Encodes* **necessários** dos dados. Lembre-se de verificar se de fato todos os dados categóricos precisam dessa transformação.

In [136]:
print(df["Quality"].value_counts())
print(df["Unit"].value_counts())


Refurbish    176
Reject       156
Accept       149
Name: Quality, dtype: int64
SP    223
BH    146
RJ    112
Name: Unit, dtype: int64


In [137]:

def encode_quality(quality):
  """
  Codifica a qualidade em um valor binário.

  Args:
    quality: O valor da qualidade.

  Returns:
    O valor codificado da qualidade.
  """

  quality_codes = {
    "Reject": 1,
    "Refurbish": 1,
    "Accept": 1
  }

  return quality_codes[quality]


df["Quality_Reject"] = df["Quality"].map(encode_quality)
df["Quality_Refurbish"] = df["Quality"].map(encode_quality)
df["Quality_Accept"] = df["Quality"].map(encode_quality)

# Realiza o hot encode da coluna Unit
df["Unit_SP"] = df["Unit"].apply(lambda x: 1 if x == "SP" else 0)
df["Unit_RJ"] = df["Unit"].apply(lambda x: 1 if x == "RJ" else 0)
df["Unit_BH"] = df["Unit"].apply(lambda x: 1 if x == "BH" else 0)

# Realiza o hot encode da coluna Quality
#df["Quality_Reject"] = df["Quality"].apply(lambda x: 1 if x == "Reject" else 0)
#df["Quality_Refurbish"] = df["Quality"].apply(lambda x: 1 if x == "Refurbish" else 0)
#df["Quality_Accept"] = df["Quality"].apply(lambda x: 1 if x == "Accept" else 0)

# Realiza o hot encode da coluna Unit
#df["Unit_SP"] = df["Unit"].apply(lambda x: 1 if x == "SP" else 0)
#df["Unit_RJ"] = df["Unit"].apply(lambda x: 1 if x == "RJ" else 0)
#df["Unit_BH"] = df["Unit"].apply(lambda x: 1 if x == "BH" else 0)
#df = pd.get_dummies(df, columns=["Quality", "Unit"])

display(df)

id    A    B    C    D    Quality Unit  Quality_Reject  \
0    559  4.9  3.1  1.5  0.1     Reject   SP               1   
1    629  4.8  3.4  1.6  0.2     Reject   SP               1   
2    192  6.7  2.5  5.8  1.8  Refurbish   RJ               1   
3    359  7.6  3.0  6.6  2.1  Refurbish   RJ               1   
4      9  4.9  3.1  1.5  0.1     Reject   RJ               1   
..   ...  ...  ...  ...  ...        ...  ...             ...   
495  155  4.6  3.6  1.0  0.2     Reject   SP               1   
496  120  6.9  3.2  5.7  2.3  Refurbish   RJ               1   
497  321  6.7  3.1  5.6  2.4  Refurbish   RJ               1   
498  448  5.6  3.0  4.1  1.3     Accept   SP               1   
499  290  4.4  2.9  1.4  0.2     Reject   BH               1   

     Quality_Refurbish  Quality_Accept  Unit_SP  Unit_RJ  Unit_BH  
0                    1               1        1        0        0  
1                    1               1        1        0        0  
2                    1               1        0        1        0  
3                    1               1        0        1        0  
4                    1               1        0        1        0  
..                 ...             ...      ...      ...      ...  
495                  1               1        1        0        0  
496                  1               1        0        1        0  
497                  1               1        0        1        0  
498                  1               1        1        0        0  
499                  1               1        0        0        1  

[481 rows x 13 columns]

In [ ]:
import pandas as pd

# Retirar os dados repetidos
columns_to_check = ['Quality_Reject', 'Quality_Refurbish', 'Quality_Accept', 'Unit_SP', 'Unit_RJ', 'Unit_BH']

# Lista para armazenar os índices das linhas duplicadas
duplicate_indices = []

# Loop para comparar cada linha com as demais
for index, row in df.iterrows():
    subset = df.drop(index)  # Remove a linha atual do DataFrame
    if (subset[columns_to_check] == row[columns_to_check]).all(axis=1).any():
        duplicate_indices.append(index)

# Crie um novo DataFrame sem as linhas duplicadas
df_without_duplicates = df.drop(duplicate_indices)

# Imprime o DataFrame sem as linhas duplicadas
print(df_without_duplicates)

print(df)

In [122]:
#Elimina linha com dados repetidos
df.drop(2, inplace=True)
display(df)


id    A    B    C    D    Quality Unit  Quality_Reject  \
0    559  4.9  3.1  1.5  0.1     Reject   SP               1   
1    629  4.8  3.4  1.6  0.2     Reject   SP               1   
3    359  7.6  3.0  6.6  2.1  Refurbish   RJ               1   
4      9  4.9  3.1  1.5  0.1     Reject   RJ               1   
5    277  7.2  3.6  6.1  2.5  Refurbish   SP               1   
..   ...  ...  ...  ...  ...        ...  ...             ...   
495  155  4.6  3.6  1.0  0.2     Reject   SP               1   
496  120  6.9  3.2  5.7  2.3  Refurbish   RJ               1   
497  321  6.7  3.1  5.6  2.4  Refurbish   RJ               1   
498  448  5.6  3.0  4.1  1.3     Accept   SP               1   
499  290  4.4  2.9  1.4  0.2     Reject   BH               1   

     Quality_Refurbish  Quality_Accept  Unit_SP  Unit_RJ  Unit_BH  
0                    1               1        1        0        0  
1                    1               1        1        0        0  
3                    1               1        0        1        0  
4                    1               1        0        1        0  
5                    1               1        1        0        0  
..                 ...             ...      ...      ...      ...  
495                  1               1        1        0        0  
496                  1               1        0        1        0  
497                  1               1        0        1        0  
498                  1               1        1        0        0  
499                  1               1        0        0        1  

[480 rows x 13 columns]

### Normalize os dados

Normalize os dados com o `StandardScaler` (melhor) ou a função `scale`.

$$ z_i = \frac{x_i - \bar{x}}{\sigma(x)}$$

E não esqueça de excluir dados não úteis ao treinamento. Depois de normalizar os dados os valores devem apresentar média próxima de 0 e desvio padrão próximo de 1.

In [138]:
# seu código
# normalizando dados de entrada
from sklearn.preprocessing import StandardScaler


# Carregue seus dados em um DataFrame, excluindo colunas não úteis
df1 = df[['A', 'B','C','D','Unit_SP','Unit_RJ','Unit_BH']]

# Cria um objeto StandardScaler
scaler = StandardScaler()

# Ajusta o StandardScaler aos dados
scaler.fit(df1)

# Normaliza os dados
X = scaler.transform(df1)

# Exibe as primeiras linhas dos dados normalizados
print(X[0:5])
print(df1.describe())


[[-1.14882737  0.06380163 -1.31316415 -1.4535803   1.07561642 -0.55092925
  -0.66016732]
 [-1.27154244  0.75085942 -1.25633205 -1.32448648  1.07561642 -0.55092925
  -0.66016732]
 [ 1.06004391 -1.31031397  1.13061614  0.74101462 -0.92969946  1.81511511
  -0.66016732]
 [ 2.16447954 -0.16521764  1.58527294  1.12829608 -0.92969946  1.81511511
  -0.66016732]
 [-1.14882737  0.06380163 -1.31316415 -1.4535803  -0.92969946  1.81511511
  -0.66016732]]
                A           B           C           D     Unit_SP     Unit_RJ  \
count  481.000000  481.000000  481.000000  481.000000  481.000000  481.000000   
mean     5.836175    3.072141    3.810603    1.225988    0.463617    0.232848   
std      0.815744    0.437099    1.761401    0.775437    0.499194    0.423086   
min      4.300000    2.000000    1.000000    0.100000    0.000000    0.000000   
25%      5.100000    2.800000    1.500000    0.300000    0.000000    0.000000   
50%      5.900000    3.000000    4.400000    1.300000    0.000000   

(Moodle) **Q1**. Após todas as transformações a soma dos valores absolutos ( `np.abs()` ) dos atributos referentes ao hot encode dos *BH*, *RJ* e *SP* são respectivamente?

*Empregue esta pergunta como ponto de checagem das suas transformações. Se não encontrar os seus valores dentre as alternativas há provavelmente um erro e você deve revisar as transformações.*


SEI QUE NÃO TEM GABARITO PARA ESSES VALORES MAS NÃO CONSIGO ENXERGA DE OUTRA FORMA .
 A soma dos valores absolutos:
BH: 146
RJ: 112
SP: 223

In [139]:
# seu código


# Obtém os valores dos atributos referentes ao hot encode dos BH, RJ e SP
bh = df['Unit_BH']
rj = df['Unit_RJ']
sp = df['Unit_SP']

# Calcula a soma dos valores absolutos dos atributos
bh_abs = np.abs(bh).sum()
rj_abs = np.abs(rj).sum()
sp_abs = np.abs(sp).sum()

print('\n A soma dos valores absolutos: ')

# Imprime os resultados
print("BH:", bh_abs)
print("RJ:", rj_abs)
print("SP:", sp_abs)


 A soma dos valores absolutos: 
BH: 146
RJ: 112
SP: 223


### Conjuntos de Treinamento e Teste

Separe os dados de treinamento e teste (30%, estratificados pela variável objetivo e com random_state = 1984). Não empregue outros parâmetros não solicitados.


In [140]:
import sklearn.model_selection

# Define a variável y
y = df["Quality_Reject"]

# Separa os dados de treinamento e teste
X_train, X_test, y_train, y_test = sklearn.model_selection.train_test_split(
    X, y, test_size=0.3, stratify=y, random_state=1984
)

# Imprime os shapes dos conjuntos de dados
print(X_train.shape, X_test.shape, y_train.shape, y_test.shape)

(336, 7) (145, 7) (336,) (145,)


# K-Vizinhos mais Próximos

Empregue o `GridSearchCV` com 5 partições e `scoring='accuracy'` para encontrar o melhor modelo entre os valores de $k$ de $4$ a $11$ e empregando as métricas euclidiana e manhattan. Verifique o modelo obtidos e o `classification_report` desse modelo.

Você pode querer empregar o modelo de código no final desse notebook que faz a mesma construção para Árvores de Decisão.

In [141]:
from sklearn.neighbors import KNeighborsClassifier
KNeighborsClassifier().get_params()

{'algorithm': 'auto',
 'leaf_size': 30,
 'metric': 'minkowski',
 'metric_params': None,
 'n_jobs': None,
 'n_neighbors': 5,
 'p': 2,
 'weights': 'uniform'}

In [142]:


# Define a variável y
y = df["Quality_Reject"]

# Cria o modelo base
knn = KNeighborsClassifier()

# Define o grid de parâmetros
param_grid = {
    "n_neighbors": list(range(4, 12)),
    "metric": ["euclidean", "manhattan"],
}

# Cria o GridSearchCV
clf = GridSearchCV(knn, param_grid, scoring="accuracy", cv=5)

# Treina o modelo
clf.fit(X_train, y_train)

# Imprime o melhor modelo
print(clf.best_estimator_)

# Imprime o relatório de classificação
y_pred = clf.predict(X_test)
print(classification_report(y_test, y_pred))

KNeighborsClassifier(metric='euclidean', n_neighbors=4)
              precision    recall  f1-score   support

           1       1.00      1.00      1.00       145

    accuracy                           1.00       145
   macro avg       1.00      1.00      1.00       145
weighted avg       1.00      1.00      1.00       145



[link text](https://)(Moodle) **Q2** Qual o melhor modelo de Knn obtido e qual a sua acuracidade?


Pela saída do classification_report, o melhor modelo de KNN obtido usa a métrica 'euclidean' e k=4. A acurácia desse modelo é 1.00, o que significa que ele classificou corretamente todas as amostras do conjunto de teste. Portanto, o melhor modelo é o KNN com métrica euclidiana e k=4, e sua acurácia é 1.00, ou seja, 100%.


(Moodle) **Q3** Qual classe teve os elementos mais erroneamente classificados?

 Parece haver elementos erroneamente classificados para a classe "1".

# Regressão Logística

Empregue o `GridSearchCV` com 5 partições e `scoring='accuracy'` para avaliar o modelo de regressão logística. Empregue somente os parâmetros padrão do `sciki-learn` para regressão logística, para isso basta empregar `param_grid = {}`. Verifique os resultados obtidos desse modelo com o `classification_report`.

Você pode querer empregar o modelo de código no final desse notebook que faz a mesma construção para Árvores de Decisão.

In [143]:
from sklearn.linear_model import LogisticRegression
LogisticRegression().get_params()

{'C': 1.0,
 'class_weight': None,
 'dual': False,
 'fit_intercept': True,
 'intercept_scaling': 1,
 'l1_ratio': None,
 'max_iter': 100,
 'multi_class': 'auto',
 'n_jobs': None,
 'penalty': 'l2',
 'random_state': None,
 'solver': 'lbfgs',
 'tol': 0.0001,
 'verbose': 0,
 'warm_start': False}

In [144]:
# seu código
# Define a variável y
y = df["Quality_Reject"]

# Cria o modelo base
logistic = LogisticRegression()

# Cria o GridSearchCV
clf = GridSearchCV(logistic, param_grid={}, scoring="accuracy", cv=5)

# Treina o modelo
clf.fit(X_train, y_train)

# Imprime o melhor modelo
print(clf.best_estimator_)

# Imprime o relatório de classificação
y_pred = clf.predict(X_test)
print(classification_report(y_test, y_pred))


ValueError: ignored

(Moodle) **Q4** Qual a acuracidade do modelo de regressão logística e qual classe teve mais falsos positivos?




A acurácia do modelo de regressão logística é de 79%.

A classe 1 (rejeitar) teve mais falsos positivos.


# Aplicando o melhor Modelo

Empregue o melhor modelo obtido acima para estimar a qualidade das peças das 10 peças do conjunto abaixo.

In [109]:
df_new = pd.read_csv('http://meusite.mackenzie.br/rogerio/TIC/pieces_new.csv')
df_new

id    A    B    C    D Unit
0  182  5.5  2.6  4.4  1.2   SP
1  345  5.8  2.7  4.1  1.0   SP
2   42  5.1  3.5  1.4  0.2   SP
3   37  6.3  2.5  4.9  1.5   BH
4   61  6.0  3.0  4.8  1.8   RJ

## Preparação dos Dados

Lembre-se, as mesmas operações feitas com os dados de treinamento precisarão ser feitas aqui também.

### Seleção de Atributos

In [110]:
import pandas as pd
from sklearn.linear_model import LogisticRegression
from sklearn.preprocessing import StandardScaler
from sklearn.model_selection import train_test_split
from sklearn.metrics import classification_report

In [ ]:
# seu código
# Selecionando os atributos
df_new = df_new.drop(columns=['id'])

### Hot encode  

In [111]:
# Fitando o encoder
hot_encode = hot_encode.fit(df_new[['Unit']])

# Transformando os dados
df_new = pd.concat(
    [df_new.drop(columns='Unit'), pd.DataFrame(hot_encode.transform(df_new[['Unit']]).toarray(), columns=list(hot_encode.categories_[0]))], axis=1,)

# Inicializando o escalador
scaler = StandardScaler()

# Fitando o escalador
scaler = scaler.fit(df_new.drop(columns='Quality_Reject'))

# Transformando os dados
df_new[['A', 'B', 'C', 'D']] = scaler.transform(df_new[['A', 'B', 'C', 'D']])

# Carregando o modelo
model = LogisticRegression()

# Carregando os dados de treinamento
X_train = df_new.drop(columns='Quality_Reject')
y_train = df_new['Quality_Reject']

# Treinando o modelo
model.fit(X_train, y_train)


# Estimando a qualidade das peças
y_pred = model.predict(df_new.drop(columns='Quality_Reject'))

# Imprimindo o relatório de classificação
print(classification_report(y_true=df_new['Quality_Reject'], y_pred=y_pred))




KeyError: ignored

### Normalize os dados

In [ ]:
# seu código

X_new = ...

X_new

## Predição

In [ ]:
# seu código

from sklearn.linear_model import LogisticRegression
from sklearn.metrics import classification_report

clf = ...

clf.fit(...)

y_pred = ...

df_new['Prediction'] = ...

df_new

(Moodle) **Q5** Quais as classes obtidas com o melhor modelo para os cinco novos casos?

# Apêndice: Árvore de Decisão (RESOLVIDO)

Este é apenas um exemplo empregando o `GridSearchCV` com 5 partições e `scoring='accuracy'` para avaliar um modelo de Árvore de Decisão com diferentes parâmetros. Você irá estudar mais detalhtes do modelo de Árvore de Decisão na próxima trilha. Mas ele é um classificador, assim como o modelo de K-Vizinhos mais Próximos e a Regressão Logística, e o uso desse estimador segue os mesmos moldes desses estimadores. Você pode assim aplicá-lo do mesmo modo.

In [92]:
from sklearn.tree import DecisionTreeClassifier
DecisionTreeClassifier().get_params()

{'ccp_alpha': 0.0,
 'class_weight': None,
 'criterion': 'gini',
 'max_depth': None,
 'max_features': None,
 'max_leaf_nodes': None,
 'min_impurity_decrease': 0.0,
 'min_samples_leaf': 1,
 'min_samples_split': 2,
 'min_weight_fraction_leaf': 0.0,
 'random_state': None,
 'splitter': 'best'}

In [93]:
from sklearn.tree import DecisionTreeClassifier

from sklearn.model_selection import GridSearchCV
from sklearn.metrics import classification_report

base_estimator = DecisionTreeClassifier()

param_grid = {'criterion': ['gini', 'entropy'], 'max_depth': [ None, 3 ]}

clf = GridSearchCV(base_estimator, param_grid, cv=5, scoring='accuracy')
clf.fit(X_train, y_train)

# print(clf.cv_results_)
print(clf.best_estimator_)

print()
print("Detailed classification report:")
print()
y_pred = clf.predict(X_test)
print(classification_report(y_test, y_pred))
print()

ValueError: ignored